pkg/controller/servicemesh/memberroll/controller.go

```diff
@@ -482,52 +483,75 @@ func (r *MemberRollReconciler) reconcileNamespaces(ctx context.Context, namespac
 		return nil, nil, err, nil
 	}
 
-	for ns := range namespacesToRemove {
-		if ns == smmr.Namespace {
-			// we never operate on the control plane namespace
-			continue
-		}
-		err = reconciler.removeNamespaceFromMesh(ctx, ns)
-		if err != nil {
-			nsErrors = append(nsErrors, err)
-			pending.Insert(ns) // TODO: do we want to have a separate list for namespaces that are being removed, but haven't yet been?
-			r.EventRecorder.Event(
-				smmr,
-				corev1.EventTypeWarning,
-				string(eventReasonErrorRemovingNamespace),
-				fmt.Sprintf("Error removing namespace %q from mesh: %s", ns, err))
-		} else {
-			configured.Delete(ns)
-		}
+	reqLogger.Info("Reconciling member namespaces")
+	startTime := time.Now()
+
+	namespacesToProcess := namespacesToReconcile.Union(namespacesToRemove)
+	if namespacesToProcess.Has(smmr.Namespace) {
+		// we never operate on the control plane namespace
+		reqLogger.Info("ignoring control plane namespace in members list of ServiceMeshMemberRoll")
+		namespacesToProcess.Delete(smmr.Namespace)
+	}
+
+	// use scatter-gather pattern to process namespaces concurrently
+	type result struct {
+		ns string
+		err error
+	}
+	results := make(chan result, len(namespacesToProcess))
+
+	// scatter
+	for namespace := range namespacesToProcess {
+		go func(ns string) {
+			var err error
+			if namespacesToReconcile.Has(ns) {
+				err = reconciler.reconcileNamespaceInMesh(ctx, ns)
+				if err != nil {
+					if errors.IsNotFound(err) || errors.IsGone(err) { // TODO: this check should be performed inside reconcileNamespaceInMesh
+						reqLogger.Info("namespace to configure with mesh is missing", "namespace", ns)
+						r.EventRecorder.Event(
```

Comment:
should we maybe record events in the outer for loop that processes results? I think it is thread-safe, so we can do it right here, I'm just wondering

Reply:
I actually had it like that in the first version, but I think it's more semantically correct for the events to be recorded as soon as they occur. I mean, an Event object represents an actual event that has happened in the system. This implies that it should be recorded as soon as the actual event happens, not seconds later (or even minutes, as could theoretically happen in this case).